R/epiparam.R

```diff
@@ -0,0 +1,304 @@
+}
+
+#' Print method for epiparam class
+#'
+#' @param x epiparam object
+#' @param ... further arguments passed to or from other methods
+#'
+#' @return Nothing (prints output)
+#' @export
+#'
+#' @examples
+#' epiparam <- epiparam()
+#' epiparam
+print.epiparam <- function(x, ...) {
+  format(x, ...)
+}
+
+#' Format method for epiparam class
+#'
+#' @param x epiparam object
+#' @param ... further arguments passed to or from other methods
+#'
+#' @return Nothing (prints output)
+#' @export
+#'
+#' @examples
+#' x <- epiparam()
+#' format(x)
+format.epiparam <- function(x, ...) {
+  validate_epiparam(x)
+  summ <- summary(x)
+  writeLines(
+    c(
+      sprintf("Epiparam object"),
+      sprintf("Number of distributions in library: %s", summ$num_entries),
+      sprintf("Number of diseases: %s", summ$num_diseases),
+      sprintf("Number of delay distributions: %s", summ$num_delay_dist),
+      sprintf("Number of offspring distributions: %s", summ$num_offspring_dist),
+      sprintf("Number of studies in library: %s", summ$num_studies),
+      sprintf("<Head of library>")
+    )
+  )
+  print(head(x)[c("disease", "epi_distribution", "prob_distribution")])
+  remaining_rows <- max(0, nrow(x) - 6)
+  remaining_cols <- max(0, ncol(x) - 3)
+  writeLines(
+    sprintf(
+      "<%s more rows & %s more cols not shown>",
+      remaining_rows,
+      remaining_cols
+    )
+  )
+  invisible(x)
+}
+
+#' Checks whether the object is an `epiparam`
+#'
+#' @param x An R object
+#'
+#' @return A boolean logical, `TRUE` if the object is an `epiparam` and `FALSE`
+#' if not
+#' @export
+#'
+#' @examples
+#' eparam <- epiparam()
+#'
+#' is_epiparam(eparam)
+is_epiparam <- function(x) {
+  inherits(x, "epiparam")
+}
+
+#' Summary method for epiparam class
+#'
+#' @param object epiparam object
+#' @param ... further arguments passed to or from other methods
+#'
+#' @return data frame of information
+#' @export
+#'
+#' @examples
+#' x <- epiparam()
+#' summary(x)
+summary.epiparam <- function(object, ...) {
+  num_entries <- nrow(object)
+  num_diseases <- length(unique(object$disease))
+  num_delay_dist <- sum(
+    object$epi_distribution %in% c(
+      "incubation_period", "generation_time",
+      "serial_interval", "onset_to_death")
+  )
+  num_offspring_dist <- sum(
+    object$epi_distribution %in% "offspring_distribution"
+  )
+  num_studies <- length(unique(object$DOI))
+  num_cont_dist <- nrow(object) - sum(object$discretised)
+  num_disc_dist <- sum(object$discetised)
+  num_vector_borne <- sum(object$extrinsic)
+  # return epiparam summary
+  list(num_entries = num_entries,
+       num_diseases = num_diseases,
+       num_delay_dist = num_delay_dist,
+       num_offspring_dist = num_offspring_dist,
+       num_studies = num_studies,
+       num_continuous_distributions = num_cont_dist,
+       num_discrete_distributions = num_disc_dist,
+       num_vector_borne_diseases = num_vector_borne
+  )
+}
+
+#' head method for epiparam class
```

Comment:
When I removed `head()` and `tail()` the print methods failed on [L208](https://github.com/epiverse-trace/epiparameter/blob/epiparams/R/epiparam.R#L208), because of a subsetting issue. I'm not sure why this is happening, but I remember before adding the `head()` and `tail()` methods it was not dispatching the `data.frame` method as expected, which was the original reason I added these functions. If there is a better way to set this up (perhaps using `NextMethod()` somewhere) in order to remove `head()` and `tail()` please let me know.

Reply:
Chipping in since I adapted this for the revised implementation of `scenario`, you could copy the implementation of `head()` to L208, so that you convert from `epiparam` to `data.frame`, on which you should be able to safely call `head()` with subsetting.

Reply:
But there would still be a `head()` `epiparam` method implemented and it would break the [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) rule, so I don't see the difference. Am I missing something?

Reply:
No, you can remove the `head()` implementation if you update L208 to be `print(utils::head(as.data.frame(x))[c("disease", "epi_distribution", "prob_distribution")])`.

Reply:
Yes, but before I added `head.epiparam()` the `head()` generic was not dispatching `head.data.frame()`, and instead I was getting the first 6 rows the database printed using `print.epiparam()`. Instead I want the `head.epiparam()` to print as though it were a data frame.

I'm happy to take a look at `{scenarios}` and discuss another way to implement.

Reply:
If I understand correctly your problem, everything is working as designed.

The `head()` method from `matrix`, `data.frame` and `array` will return an object of the same class as `x`.

So if you're calling `head.data.frame()` on an `epiparam` object, you get an `epiparam` object back, and it's then dispatched to `print.epiparam()` when you print it.

Reply:
Unclassing your object while calling `head()` or `tail()` seems like a potential source of errors or confusion for your users.

Reply:
I agree. I wanted `head()` and `tail()` to print all columns which is not the case for `print.epiparam()`. If you think it is bad practise to use `as.data.frame()` in the `head()` and `tail()` methods I could rename the functions to make clear that they will return a data frame and not use the generic.

Reply:
I think the issue is that the `print()` method relies on the `format()` method, which calls `print(head(...))`. So `head()` returns an `epiparam` (or in my case, `scenario`), and then looks for a `print()` method, which in turn calls `head()` and so on. Calling `NextMethod(generic = "print", object = x)` does't really seem to work

Reply:
> I agree. I wanted head() and tail() to print all columns which is not the case for print.epiparam(). If you think it is bad practise to use as.data.frame() in the head() and tail() methods I could rename the functions to make clear that they will return a data frame and not use the generic.

I'm not sure. The problem sounds more like you're not happy with the way your `print()` works. If you like the output is some situations but not in others, you may not be looking for `print()`. If you don't like it for the output of `head()`, chances are you won't like it in some other legitimate situations. Maybe what you put in `print.epiparam()` should really be `summary.epiparam()` and you should stick with `print.data.frame()` for printing.

Reply:
I like the current `print.epiparam()` implementation (which uses `summary.epiparam()`) as I did not want all 56 columns of the `epiparam` object printed. One option would be to use `tibble` to provide a cleaner print method but I didn't want the dependency. For now I'll keep with the print method as it's minimal and informative. If you both think it's best to have the `head()` and `tail()` print as `epiparam` by default then I'm happy to change and I will write two new functions that do what the current `head()` and `tail()` are doing.